oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrant.java

```diff
@@ -41,54 +43,33 @@
  */
 abstract class AbstractOAuth2AuthorizationGrant implements OAuth2AuthorizationGrant {
 
+    @SuppressWarnings("rawtypes")
+    private static final AtomicReferenceFieldUpdater<
+            AbstractOAuth2AuthorizationGrant, CompletableFuture> tokenFutureUpdater =
+            AtomicReferenceFieldUpdater.newUpdater(
+                    AbstractOAuth2AuthorizationGrant.class, CompletableFuture.class, "tokenFuture");
+
     private final RefreshAccessTokenRequest refreshRequest;
 
     private final Duration refreshBefore;
 
-    /**
-     * Holds a token object and facilitates its lifecycle.
-     */
-    private final TokenLifecycleManager<GrantedOAuth2AccessToken> tokenManager;
-
-    AbstractOAuth2AuthorizationGrant(RefreshAccessTokenRequest refreshRequest, Duration refreshBefore,
-                                     @Nullable Supplier<? extends GrantedOAuth2AccessToken> tokenSupplier,
-                                     @Nullable Consumer<? super GrantedOAuth2AccessToken> tokenConsumer,
-                                     @Nullable Executor executor) {
-        this.refreshRequest = requireNonNull(refreshRequest, "refreshRequest");
-        this.refreshBefore = requireNonNull(refreshBefore, "refreshBefore");
-        tokenManager = new TokenLifecycleManager<>(this::isValid, AbstractOAuth2AuthorizationGrant::canRefresh,
-                                                   AbstractOAuth2AuthorizationGrant::shallObtainInsteadOfUpdate,
-                                                   this::obtainAccessToken, this::refreshAccessToken,
-                                                   tokenSupplier, tokenConsumer, executor);
-    }
-
-    /**
-     * Tests the token for validity at the given {@link Instant} time.
-     */
-    private boolean isValid(GrantedOAuth2AccessToken token, Instant now) {
-        return token.isValid(now.plus(refreshBefore));
-    }
+    @Nullable
+    private final Supplier<CompletableFuture<? extends GrantedOAuth2AccessToken>> loadTokenFunc;
 
-    /**
-     * Tests whether the token object can be refreshed or re-obtained.
-     */
-    private static boolean canRefresh(GrantedOAuth2AccessToken token) {
-        return token.refreshToken() != null;
-    }
+    @Nullable
+    private final Consumer<? super GrantedOAuth2AccessToken> saveTokenFunc;
```

Comment:
They are replacements for `tokenPersistencySupplier` and `tokenPersistencyConsumer` in the original implementation  so the names.

And I agree that it is awkward having persistent layer logic in this class and I am considering removing those completely as I've stated here(https://github.com/line/armeria/pull/3618#issuecomment-872854532).

It would be nice if I can get some advice!

Reply:
Renamed them!